src/packet/decode.ts

```diff
@@ -0,0 +1,111 @@
+import * as RLP from "rlp";
+import {
+  IAuthMessagePacket,
+  IMessagePacket,
+  IWhoAreYouPacket,
+  Magic,
+  Packet,
+  PacketType,
+  Tag,
+} from "./types";
+
+import {
+  AUTH_TAG_LENGTH,
+  ERR_INVALID_BYTE_SIZE,
+  ERR_TOO_LARGE,
+  ERR_TOO_SMALL,
+  ERR_UNKNOWN_FORMAT,
+  MAX_PACKET_SIZE,
+  ID_NONCE_LENGTH,
+  TAG_LENGTH,
+} from "./constants";
+
+/**
+ * Decode raw bytes into a packet. The `magic` value (SHA2256(node-id, b"WHOAREYOU")) is passed as a parameter to check
+ * for the magic byte sequence.
+ */
+export function decode(data: Buffer, magic: Magic): [PacketType, Packet] {
+  if (data.length > MAX_PACKET_SIZE) {
+    throw new Error(ERR_TOO_LARGE);
+  }
+  // ensure the packet is large enough to contain the correct headers
+  if (data.length < TAG_LENGTH + AUTH_TAG_LENGTH + 1) {
+    throw new Error(ERR_TOO_SMALL);
+  }
+  const tag = data.slice(0, TAG_LENGTH);
+  data = data.slice(TAG_LENGTH);
```

Comment:
I would leave comment on method that data input will be modified or create new buffer from input(performance issue maybe)